src/lib/telemetry/TelemetryReporter/PendingTelemetryDataManager.ts

```diff
@@ -0,0 +1,142 @@
+import { TelemetryCollector, TelemetryFile } from '../TelemetryCollector'
+
+export class PendingTelemetryDataManager {
+  public static readonly PENDING_DATA_DIR = join(TelemetryCollector.TELEMETRY_LOCAL_DIR, 'pendingData')
+  private static readonly MAX_TELEMETRY_DIR_SIZE = 10 * 1000 * 1000
+  private static readonly MAX_TELEMETRY_DIR_NUMBER_OF_FILES = 500
+
+  private static singleton: PendingTelemetryDataManager
+  public static getSingleton() {
+    if (!PendingTelemetryDataManager.singleton) {
+      PendingTelemetryDataManager.singleton = new PendingTelemetryDataManager(
+        PendingTelemetryDataManager.PENDING_DATA_DIR
+      )
+    }
+
+    return PendingTelemetryDataManager.singleton
+  }
+
+  private pendingTelemetryFilesToCreate: TelemetryFile[]
+  private pendingMetaMetrics: MetricReport
+  private dataPendingLock: FileLock
+
+  constructor(private pendingDataDir: string) {
+    const dataPendingLockName = 'reporter.lock'
+    const dataPendingLockPath = join(pendingDataDir, dataPendingLockName)
+    this.dataPendingLock = new FileLock(dataPendingLockPath)
+
+    this.pendingTelemetryFilesToCreate = []
+    this.pendingMetaMetrics = MetricReport.create({
+      command: 'not-applicable',
+    })
+  }
+
+  public async acquireLock() {
+    const getLockTime = process.hrtime()
+    await this.dataPendingLock.lock()
+    this.registerPendingMetaMetric(TelemetryMetaMetrics.PENDING_DATA_ACQUIRE_LOCK_TIME, process.hrtime(getLockTime))
+  }
+
+  public releaseLock() {
+    return this.dataPendingLock.unlock()
+  }
+
+  public async ensureTelemetryDirMaxSize() {
+    const { totalDirSize, numberOfFiles } = await this.getPendingDirStats()
+
+    if (
+      totalDirSize > PendingTelemetryDataManager.MAX_TELEMETRY_DIR_SIZE ||
+      numberOfFiles > PendingTelemetryDataManager.MAX_TELEMETRY_DIR_NUMBER_OF_FILES
+    ) {
+      await remove(TelemetryCollector.TELEMETRY_LOCAL_DIR)
+    }
+  }
+
+  public async getFilePaths() {
+    await ensureDir(this.pendingDataDir)
+    const pendingFiles = (await readdir(this.pendingDataDir)).filter(
+      fileName => fileName !== this.dataPendingLock.lockName
+    )
+
+    return pendingFiles.map(pendingFile => join(this.pendingDataDir, pendingFile))
+  }
+
+  public async moveTelemetryFileToPending(filePath: string) {
+    return move(filePath, this.newPendingFilePath())
+  }
+
+  public registerPendingFile(content: TelemetryFile) {
+    this.pendingTelemetryFilesToCreate.push(content)
+  }
+
+  public registerPendingMetaMetric(metricName: string, latency: [number, number]) {
+    this.pendingMetaMetrics.addMetric(metricName, hrTimeToMs(latency))
+  }
+
+  public async createPendingFiles() {
+    await ensureDir(this.pendingDataDir)
+    this.flushPendingMetaMetrics()
+
+    await Promise.all(
+      this.pendingTelemetryFilesToCreate.map(async fileContent => {
+        try {
+          await writeJson(this.newPendingFilePath(), fileContent)
+        } catch (err) {
+          console.error(err)
```

Comment:
Will console.log work inside report script?

Reply:
Unfortunately no, but I couldn't think on a way to report these errors. If we get errors while creating pending errors at this point I thought we couldn't do much

Reply:
Couldn't we keep the catch block empty then? Just to avoid wrong expectations from someone that look into the code and see the log.

Reply:
Ok!